build: Enable React Compiler for Browserify builds, fix `react-compiler/react-compiler` ESLint rule violations

MajorLift opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## Motivation

This commit enables React Compiler for Babel-transformed Browserify builds, and for ESLint (via the `react-compiler/react-compiler` ruleset). We can expect long-term UI performance benefits and developer time savings from this change. The Compiler transparently adds memoization code for React components and hooks, among other optimizations, and enforces React best practices.

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Because enabling React Compiler requires addressing all violations of the `react-compiler/react-compiler` rule, this commit contains a large number of formatting changes and renames that do not require close review, alongside refactors and logical changes that do. 

I would suggest that reviewers primarily focus on whether a given change is problematic, rather than on whether it is necessary. If the motivation for any change is unclear, it's safe to assume that it fixes an error thrown in CI. The objective of this commit is to ship fixes that unblock React Compiler without introducing regressions in functionality or performance. Suggestions on implementing React best practices will be catalogued and addressed in follow-up tickets.

The 10% bundle size increase is due to new memoizations being applied as React Compiler transforms files during the build process, which is the tradeoff for improvements in runtime interaction and responsiveness metrics. Despite this change, the UI startup and pageLoad benchmark results show no evidence of degradations in initial load time.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37480?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null (non-user facing but regressions possible)

## **Related issues**

- Epic: https://github.com/MetaMask/metamask-extension/issues/33059
- Previous: https://github.com/MetaMask/metamask-extension/pull/37383
- Next: https://github.com/MetaMask/metamask-extension/pull/38007

## **Manual testing steps**

1. Create a Browserify dev build by running `yarn start`.
2. Inspect `./dist/(chrome|firefox)/ui-[3-17].js`.
3. Search for "`_reactCompilerRuntime.c`".
4. There should be many examples of React Compiler-optimized code that uses a `$` variable e.g.
```js
  const $ = (0, _reactCompilerRuntime.c)(22);
  const {
    pathname
  } = (0, _reactRouterDomV5Compat.useLocation)();
  const dispatch = (0, _reactRedux.useDispatch)();
  const t = (0, _useI18nContext.useI18nContext)();
  let t0;
  if ($[0] !== pathname) {
    t0 = (0, _snaps2.decodeSnapIdFromPathname)(pathname);
    $[0] = pathname;
    $[1] = t0;
  } else {
    t0 = $[1];
  }
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.














<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables React Compiler in Browserify Babel builds and enforces the react-compiler ESLint rule, with broad hook/memoization refactors, selector renames, tests updates, and dependency/policy changes.
> 
> - **Build/Tooling**:
>   - Enable React Compiler via Babel override for `ui/**/{components,contexts,hooks,layouts,pages}` with `babel-plugin-react-compiler` (target 17).
>   - Add `react-compiler-runtime` and `babel-plugin-react-compiler` deps; update `esbuild`, `koa` (v3), and related tooling.
>   - Enforce `react-compiler/react-compiler` ESLint rule and plugin across JS/TS React configs.
> - **Policies/Config**:
>   - Extend LavaMoat policies and depcheck ignores for React Compiler packages and updated sub-deps.
> - **UI/Hook Refactors (compliance + perf)**:
>   - Replace patterns violating React Compiler (e.g., `useMemo` state updates, unstable callbacks, missing keys) with `useCallback`/`useRef`/guards; fix array checks, keys, debounce handling, and effect deps.
>   - Improve polling hooks (`usePolling`, `useMultiPolling`) with input change detection and cleanup; stabilize navigation and debounce handlers.
>   - Add `isNativeAsset` and use in `AssetPage`; compute asset balances immutably.
> - **Reliability/Null Safety**:
>   - Guard optional `request?.metadata?.id` across confirmations/snaps flows; miscellaneous null-safe checks.
> - **Selectors/Types**:
>   - Rename selectors: `useSafeChainsListValidationSelector` → `getUseSafeChainsListValidation`; `use4ByteResolutionSelector` → `getUse4ByteResolution` (and propagate).
>   - Tighten types (e.g., `ownerId` optional handling, asset type helpers).
> - **Tests**:
>   - Update tests to use `renderHook` for hooks; adjust mocks to renamed selectors and new patterns.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c47e3f6fdd604d9a5c22cd52c034dfd2e753a981. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

MajorLift committed:
Set `react-hooks/rules-of-hooks` eslint ruleset to "error"

MajorLift committed:
Fix `rules-of-hooks` violations

MajorLift committed:
Rename: `{s,useS}etConfirmationAlerts`

MajorLift committed:
Add `eslint-disable` directives for `rules-of-hooks`

MajorLift committed:
Rename: `{s,useS}yncConfirmPath`

MajorLift committed:
Set up React Compiler

MajorLift committed:
Fix formatting rule violations

MajorLift committed:
Fix logical rule violations

MajorLift committed:
Fix breakages

georgewrmarshall reviewed (dismissed):
Approving component library changes on behalf of the @MetaMask/design-system-engineers

MajorLift committed:
Merge branch 'main' into jongsun/chore/251030-enable-rules-of-hooks

MajorLift committed:
Fix conditional hook calls introduced from main

MajorLift committed:
Fix incorrectly resolved merge conflict

davidmurdoch commented on package.json:
```diff
@@ -674,6 +677,7 @@
     "process": "^0.11.10",
     "pumpify": "^2.0.1",
     "randomcolor": "^0.5.4",
+    "react-compiler-webpack": "^0.2.0",
```
pin this to exactly `0.2.0`. In semver versioning the `0` version allows for breaking changes at any time.

MajorLift committed:
Refactor webpack config to have dedicated loader for react compiler

MajorLift committed:
Fix lint regression

MajorLift committed:
Fix regression from main

MajorLift committed:
Fix webpack react compiler loader option object type

MajorLift committed:
Fix incorrectly resolved merge conflict

MajorLift commented:
@metamaskbot update-policies

MajorLift committed:
Merge branch 'main' into jongsun/chore/251030-enable-rules-of-hooks

MajorLift committed:
Merge branch 'main' into jongsun/chore/251030-enable-rules-of-hooks

MajorLift committed:
Use exact versions for react compiler deps

MajorLift committed:
Merge branch 'jongsun/chore/251030-enable-rules-of-hooks' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
More fixes

MajorLift committed:
Merge branch 'jongsun/chore/251030-enable-rules-of-hooks' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Rename `{use,handle}SnapNavigate`

MajorLift committed:
lavamoat: Widen `process` nested globals access to top-level scope

metamaskbot committed:
Update LavaMoat policies

MajorLift commented:
@metamaskbot update-policies

MajorLift committed:
Fix ui path in webpack loader

metamaskbot committed:
Update LavaMoat policies

MajorLift committed:
Remove source path exceptions

MajorLift committed:
test: Update unit test snapshot

MajorLift committed:
Exclude `typed-sign-permission-util.ts`

MajorLift committed:
Fix error from missing `t` in dep array

```
SES_UNCAUGHT_EXCEPTION: TypeError: Cannot convert a Symbol value to a string
  at Array.join (<anonymous>)
  at Array.toString (<anonymous>)
  at Array.join (<anonymous>)
  at areHookInputsEqual (react-dom.development.js:14931:1)
  at updateMemo (react-dom.development.js:15861:1)
  at Object.useMemo (react-dom.development.js:16413:1)
  at useMemo (react.development.js:1532:1)
  at useBurnAddressAlert (useBurnAddressAlert.ts:38:17)
  at useTransactionAlerts (useConfirmationAlerts.ts:41:28)
  at useConfirmationAlerts (useConfirmationAlerts.ts:98:29)
  at useSetConfirmationAlerts (useSetConfirmationAlerts.ts:14:18)
  at ConfirmAlerts (confirm-alerts.tsx:13:3)
  at renderWithHooks (react-dom.development.js:14985:1)
  at updateFunctionComponent (react-dom.development.js:17356:1)
  at beginWork (react-dom.development.js:19063:1)
  at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
  at HTMLUnknownElement.sentryWrapped (sentry-install.js:22550:17)
  at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
  at invokeGuardedCallback (react-dom.development.js:4056:1)
  at beginWork$1 (react-dom.development.js:23964:1)
  at performUnitOfWork (react-dom.development.js:22779:1)
  at workLoopSync (react-dom.development.js:22707:1)
  at renderRootSync (react-dom.development.js:22670:1)
  at performSyncWorkOnRoot (react-dom.development.js:22293:1)
  at react-dom.development.js:11327:1
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushSyncCallbackQueueImpl (react-dom.development.js:11322:1)
  at flushSyncCallbackQueue (react-dom.development.js:11309:1)
  at flushPassiveEffectsImpl (react-dom.development.js:23620:1)
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushPassiveEffects (react-dom.development.js:23447:1)
  at react-dom.development.js:23324:1
  at workLoop (scheduler.development.js:417:1)
  at flushWork (scheduler.development.js:390:1)
  at MessagePort.performWorkUntilDeadline (scheduler.development.js:157:1)
```

MajorLift committed:
Fix "Should have a queue" error by opting-out `ui/pages/confirmations/hooks/alerts/transactions`

```
SES_UNCAUGHT_EXCEPTION: Error: Should have a queue. This is likely a bug in React. Please file an issue.
  at updateReducer (react-dom.development.js:15225:1)
  at Object.useReducer (react-dom.development.js:16425:1)
  at useReducer (react.development.js:1512:1)
  at useSelectorWithStoreAndSubscription (useSelector.js:22:1)
  at useSelector (useSelector.js:139:1)
  at useFirstTimeInteractionAlert (useFirstTimeInteractionAlert.ts:25:39)
  at useTransactionAlerts (useConfirmationAlerts.ts:42:37)
  at useConfirmationAlerts (useConfirmationAlerts.ts:98:29)
  at useSetConfirmationAlerts (useSetConfirmationAlerts.ts:14:18)
  at ConfirmAlerts (confirm-alerts.tsx:13:3)
  at renderWithHooks (react-dom.development.js:14985:1)
  at updateFunctionComponent (react-dom.development.js:17356:1)
  at beginWork (react-dom.development.js:19063:1)
  at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
  at HTMLUnknownElement.sentryWrapped (sentry-install.js:22550:17)
  at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
  at invokeGuardedCallback (react-dom.development.js:4056:1)
  at beginWork$1 (react-dom.development.js:23964:1)
  at performUnitOfWork (react-dom.development.js:22779:1)
  at workLoopSync (react-dom.development.js:22707:1)
  at renderRootSync (react-dom.development.js:22670:1)
  at performSyncWorkOnRoot (react-dom.development.js:22293:1)
  at react-dom.development.js:11327:1
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushSyncCallbackQueueImpl (react-dom.development.js:11322:1)
  at flushSyncCallbackQueue (react-dom.development.js:11309:1)
  at flushPassiveEffectsImpl (react-dom.development.js:23620:1)
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushPassiveEffects (react-dom.development.js:23447:1)
  at react-dom.development.js:23324:1
  at workLoop (scheduler.development.js:417:1)
  at flushWork (scheduler.development.js:390:1)
  at MessagePort.performWorkUntilDeadline (scheduler.development.js:157:1)
```

```
That “queue” is React’s internal hook queue—the data structure React builds when it wires up useReducer, useState, useContext, etc., for a component. The warning is React telling us that the queue it created during the first render (based on the hook order it saw) no longer matches the queue it needs for a later render. React Compiler is what’s injecting extra useReducer-backed queues into hooks like useSelector; when the compiler decides a hook should be memoized, it rewrites the hook so it now manages a queue of memoized values. If that code path later needs to return “nothing” (e.g., the hook returns an empty array), the compiler-generated code still mutates the queue length. On the next render, React allocates fewer hooks than it just recorded, resulting in a mismatch.
```

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Opt-out regression from main

MajorLift committed:
Fix incorrect mounting logic

MajorLift committed:
Remove unnecessary key prop

MajorLift committed:
Fix build config

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

metamaskbot committed:
Update LavaMoat policies

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Opt-outs

MajorLift committed:
Add excluded paths

MajorLift committed:
Add react compiler logger for webpack

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

MajorLift committed:
Revert webpack config

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Remove `exhaustive-deps` regression from main

metamaskbot committed:
Update LavaMoat policies

MajorLift committed:
Remove `component-library` from excluded paths

MajorLift committed:
Remove webpack entry in `.depcheckrc`

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Fix faulty fallback logic for confirm alerts state keys

MajorLift committed:
Bump deps to v1.0

MajorLift committed:
Fix `useAsync` mounted check race condition

MajorLift committed:
Add entry to exclude patterns

metamaskbot committed:
Update LavaMoat policies

MajorLift reviewed (commented):
For @MetaMask/policy-reviewers:

MajorLift reviewed (commented):
Build system changes:

MajorLift commented on ui/selectors/selectors.js:
```diff
@@ -3498,7 +3498,7 @@ export const getUpdatedAndSortedAccountsWithCaipAccountId =
   });
 
-export const useSafeChainsListValidationSelector = (state) => {
+export const getUseSafeChainsListValidation = (state) => {
```
The `use` prefix is reserved for hooks, and selectors should begin with "get-" by convention.

Strictly speaking, this is Redux code, so it doesn't matter to React Compiler whether or not this convention is applied, and the builds aren't affected either way since `ui/selectors` is excluded. 

However, our React linter rulesets are set up to apply to `ui/**/*`, and it makes sense to enforce naming conventions consistently.

MajorLift commented on ui/contexts/metametrics.js:
```diff
@@ -235,6 +235,7 @@ export function MetaMetricsProvider({ children }) {
+  // eslint-disable-next-line react-compiler/react-compiler
   trackEventWithMethods.bufferedTrace = bufferedTrace;
   trackEventWithMethods.bufferedEndTrace = bufferedEndTrace;
   trackEventWithMethods.onboardingParentContext = onboardingParentContext;
```
Mutates data in-place, which is an anti-pattern. Opting-out because fixing causes too much downstream code to break.

MajorLift commented on ui/pages/confirmations/hooks/alerts/transactions/useFirstTimeInteractionAlert.ts:
```diff
@@ -1,3 +1,5 @@
+'use no memo';
```
The hooks in this directory have been excluded from React Compiler due to the following error. To be fixed in a follow-up ticket.

```
SES_UNCAUGHT_EXCEPTION: Error: Should have a queue. This is likely a bug in React. Please file an issue.
  at updateReducer (react-dom.development.js:15225:1)
  at Object.useReducer (react-dom.development.js:16425:1)
  at useReducer (react.development.js:1512:1)
  at useSelectorWithStoreAndSubscription (useSelector.js:22:1)
  at useSelector (useSelector.js:139:1)
  at useFirstTimeInteractionAlert (useFirstTimeInteractionAlert.ts:25:39)
  at useTransactionAlerts (useConfirmationAlerts.ts:42:37)
  at useConfirmationAlerts (useConfirmationAlerts.ts:98:29)
  at useSetConfirmationAlerts (useSetConfirmationAlerts.ts:14:18)
  at ConfirmAlerts (confirm-alerts.tsx:13:3)
  at renderWithHooks (react-dom.development.js:14985:1)
  at updateFunctionComponent (react-dom.development.js:17356:1)
  at beginWork (react-dom.development.js:19063:1)
  at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
  at HTMLUnknownElement.sentryWrapped (sentry-install.js:22550:17)
  at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
  at invokeGuardedCallback (react-dom.development.js:4056:1)
  at beginWork$1 (react-dom.development.js:23964:1)
  at performUnitOfWork (react-dom.development.js:22779:1)
  at workLoopSync (react-dom.development.js:22707:1)
  at renderRootSync (react-dom.development.js:22670:1)
  at performSyncWorkOnRoot (react-dom.development.js:22293:1)
  at react-dom.development.js:11327:1
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushSyncCallbackQueueImpl (react-dom.development.js:11322:1)
  at flushSyncCallbackQueue (react-dom.development.js:11309:1)
  at flushPassiveEffectsImpl (react-dom.development.js:23620:1)
  at unstable_runWithPriority (scheduler.development.js:468:1)
  at runWithPriority$1 (react-dom.development.js:11276:1)
  at flushPassiveEffects (react-dom.development.js:23447:1)
  at react-dom.development.js:23324:1
  at workLoop (scheduler.development.js:417:1)
  at flushWork (scheduler.development.js:390:1)
  at MessagePort.performWorkUntilDeadline (scheduler.development.js:157:1)
```

MajorLift commented on ui/hooks/useMultiPolling.ts:
The motivation for the changes in this file (and in `usePolling.ts`) are to remove `JSON.stringify()` computation from deps arrays.

The best practice fix here would be to use `useEqualityCheck` to guarantee that the hook doesn't re-render due to referential instability of `usePollingOptions.input`, but that approach breaks unit tests and causes unnecessary renders. The fix here is intended as a temporary placeholder.

MajorLift commented on ui/hooks/usePolling.ts:
```diff
@@ -13,16 +13,27 @@ const usePolling = <PollingInput>(
+  const prevPollingInputStringified = useRef<string | null>(null);
+  const hasPollingInputChanged =
+    JSON.stringify(usePollingOptions.input) !==
+    prevPollingInputStringified.current;
```
Same as above

MajorLift reviewed (commented):
Notes for reviewers:

MajorLift committed:
Add deps

MajorLift commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/hooks/useAssetMetadata.ts:
```diff
@@ -49,7 +57,7 @@ export const useAssetMetadata = (
       trimmedSearchQuery.length > 30
     ) {
-      abortControllerRef.current = new AbortController();
+      abortControllerRef.current ??= new AbortController();
```
Fixed: 94e10c81c22b412c2d6a04001e7bce949552cf2e

MajorLift commented on ui/helpers/utils/notification.util.ts:
```diff
@@ -410,7 +410,9 @@ export function hasNetworkFeeFields(
-export const getNetworkFees = async (notification: OnChainRawNotification) => {
+export const getNetworkFees = async (
+  notification: OnChainRawNotificationsWithNetworkFields,
+) => {
```
The type guard ensures runtime safety. The input type is a compile time check. They serve complementary but distinct use cases.

> this means callers could pass invalid notifications without type safety protection.

This is straight-up wrong. Invalid notifications will be caught by the type guard and will trigger the error path.

MajorLift committed:
Fix HOC-returned component local name

MajorLift commented on ui/helpers/higher-order-components/with-router-hooks/with-router-hooks.tsx:
```diff
@@ -15,7 +15,7 @@ export type RouterHooksProps = {
   WrappedComponent: React.ComponentType<Props & RouterHooksProps>,
 ): React.ComponentType<Props> {
-  const ComponentWithRouterHooks = (props: Props) => {
+  const useComponentWithRouterHooks = (props: Props) => {
```
Resolved: 66aba3d58866e39904fd458a13a28aacd4cab720

MajorLift commented on ui/components/app/qr-hardware-popover/base-reader.js:
```diff
@@ -105,30 +120,14 @@ const BaseReader = ({
       mounted.current = false;
-      clearTimeout(permissionChecker);
+      clearTimeout(permissionCheckerRef.current);
+      permissionCheckerRef.current = null;
```
- WONTFIX: Out-of-scope. We're going to preserve existing behavior.
- The dependency chain of `checkPermissions` (empty deps array) -> `initCamera` -> `checkEnvironment` -> `useEffect` is referentially stable.

MajorLift commented on ui/ducks/confirm-alerts/confirm-alerts.ts:
```diff
@@ -201,6 +201,10 @@ export default function confirmAlertsReducer(
   state: ConfirmAlertsState = INIT_STATE,
   action: Action,
 ) {
+  if (action.ownerId === undefined) {
```
- WONTFIX: Out-of-scope. We're going to preserve existing behavior.

MajorLift committed:
Opt-out `useTokenExchangeRate`

MajorLift commented on ui/components/app/currency-input/hooks/useTokenExchangeRate.tsx:
```diff
@@ -41,6 +41,19 @@ export default function useTokenExchangeRate(
+      }));
+    }
+  }, [contractExchangeRate, tokenAddress]);
+
```
This would break the unit tests if true, since they check whether the fetch call goes out, but reverting and opting-out for now: e839433dddbfdeab7aa1851f53a7e22057024414

davidmurdoch commented on .eslintrc.js:
```diff
@@ -320,8 +320,9 @@ module.exports = {
         },
       },
-      plugins: ['react'],
+      plugins: ['react', 'react-compiler'],
```
Is react-compiler the right plugin? https://react.dev/blog/2025/10/07/react-compiler-1 says

> Compiler-powered lint rules ship in eslint-plugin-react-hooks’s recommended and recommended-latest preset.

and 

> If you have already installed eslint-plugin-react-compiler, you can now remove it and use eslint-plugin-react-hooks@latest. Many thanks to [@michaelfaith](https://bsky.app/profile/michael.faith) for contributing to this improvement!

MajorLift committed:
Fix `renderHook` usage in test

MajorLift committed:
Refactor babel config to register react compiler as an override

MajorLift commented on ui/hooks/useModalProps.test.ts:
```diff
@@ -26,9 +27,11 @@ jest.mock('../store/actions', () => ({
-    expect(props).toStrictEqual(MOCK_PROPS);
-    expect(hideModal).toStrictEqual(expect.any(Function));
+      expect(props).toStrictEqual(MOCK_PROPS);
+      expect(hideModal).toStrictEqual(expect.any(Function));
```
Resolved here: https://github.com/MetaMask/metamask-extension/pull/37480/commits/8a491bcbf43aca89127d3a24c310545b90891eb2

MajorLift committed:
Dedupe lockfile

MajorLift commented:
@metamaskbot update-policies

MajorLift commented:
Running `yarn dedupe` breaks "[Update LavaMoat webpack policy](https://github.com/MetaMask/metamask-extension/actions/runs/19581006439/job/56078835834?pr=37480#logs)" with the following error:

```
Run yarn webpack --env production --no-cache --generatePolicy
/home/runner/work/metamask-extension/metamask-extension/node_modules/async-function/require.mjs:2
var __create=Object.create;var __defProp=Object.defineProperty;var __getOwnPropDesc=Object.getOwnPropertyDescriptor;var __getOwnPropNames=Object.getOwnPropertyNames;var __getProtoOf=Object.getPrototypeOf;var __hasOwnProp=Object.prototype.hasOwnProperty;var __export=(target,all)=>{for(var name in all)__defProp(target,name,{get:all[name],enumerable:true})};var __copyProps=(to,from,except,desc)=>{if(from&&typeof from==="object"||typeof from==="function"){for(let key of __getOwnPropNames(from))if(!__hasOwnProp.call(to,key)&&key!==except)__defProp(to,key,{get:()=>from[key],enumerable:!(desc=__getOwnPropDesc(from,key))||desc.enumerable})}return to};var __toESM=(mod,isNodeMode,target)=>(target=mod!=null?__create(__getProtoOf(mod)):{},__copyProps(isNodeMode||!mod||!mod.__esModule?__defProp(target,"default",{value:mod,enumerable:true}):target,mod));var __toCommonJS=mod=>__copyProps(__defProp({},"__esModule",{value:true}),mod);var require_exports={};__export(require_exports,{default:()=>require_default,"module.exports":()=>import_index.default});module.exports=__toCommonJS(require_exports);var import_index=__toESM(require("./index.js"));var require_default=import_index.default;0&&(module.exports={"module.exports"});
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

SyntaxError: Unexpected string
    at wrapSafe (node:internal/modules/cjs/loader:1692:18)
    at Module._compile (node:internal/modules/cjs/loader:1735:20)
    at Object.transformer (/home/runner/work/metamask-extension/metamask-extension/node_modules/tsx/dist/register-DCnOAxY2.cjs:2:1186)
    at Module.load (node:internal/modules/cjs/loader:1481:32)
    at Module._load (node:internal/modules/cjs/loader:1300:12)
    at TracingChannel.traceSync (node:diagnostics_channel:328:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:245:24)
    at Module.require (node:internal/modules/cjs/loader:1504:12)
    at require (node:internal/modules/helpers:152:16)
    at Object.<anonymous> (/home/runner/work/metamask-extension/metamask-extension/node_modules/get-intrinsic/index.js:154:24)

Node.js v24.11.1
```

MajorLift committed:
Revert "Dedupe lockfile"

This reverts commit e456d36eb599735270f0fef0f6e2067820ce31c4.

MajorLift committed:
Make regex non-capturing

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Fix babel config regex to only match top-level `ui/`

MajorLift committed:
Bump `tsx` from `^4.19.2` to `^4.20.6`

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

david0xd commented on ui/components/app/currency-input/hooks/useTokenExchangeRate.tsx:
```diff
@@ -66,6 +66,8 @@ export default function useTokenExchangeRate(
       contractExchangeRates[tokenAddress] || exchangeRates[tokenAddress];
 
     if (!contractExchangeRate) {
+      // TODO: Fix "Calling setState from useMemo may trigger an infinite loop"
```
Do we have some strategy for addressing these kind of things that are out of scope on this PR?

david0xd reviewed (dismissed):
Overall looks and works good! I did little bit of testing as well.

MajorLift committed:
Bump `koa` from `^2.7.0` to `^3.1.1`

metamaskbot committed:
Update LavaMoat policies

MajorLift committed:
Fix regex

davidmurdoch commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/hooks/useAssetMetadata.ts:
```diff
@@ -25,6 +26,13 @@ export const useAssetMetadata = (
+      abortControllerRef.current = null;
+    };
+  }, [abortControllerRef]);
+
```
Because `abortControllerRef` is passed in from the parent, I don't think this method should be responsible for tearing this down.

Though I also think it should just be created here in this function in the first place, and completely managed here, but there might be a good reason it isn't.

MajorLift replied:
Removed here: https://github.com/MetaMask/metamask-extension/pull/37480/commits/01094ea2b1a5e03b5fa21b9b4122caa05545ae8e

However, on second thought, I don't feel confident that the `??=` change is non-breaking.

- `useAssetMetadata` is responsible for instantiating `AbortController` instance(s) and passing their abort signals down to `fetchAssetMetadata` -> `fetchWithTimeout`.
- The parent `AssetPickerModal` is responsible for making `.abort()` calls.
- Instantiating via `??=` ensures that only one `AbortController` instance is used throughout the parent component's lifecycle, and even if multiple `fetchAssetMetadata` calls are in flight they can all be aborted from the parent component.
- This might be a change of behavior, as currently, potentially multiple long-lived fetch calls each with different `AbortController` instances could be active in closure or the task queue, but the parent component can only send abort signals to the most recently created instance.

Maybe it would be safer to opt-out this hook as well? I think the change is better for performance (not to mention the compiler optimizations), but we could follow up on this after more testing.

MajorLift committed:
Remove unnecessary cleanup

MajorLift committed:
Opt-out confirm alert files

davidmurdoch reviewed (changes requested):
requested some changes during 1:1 review

MajorLift committed:
Revert `in` property check

MajorLift committed:
Revert handling for `background[method]` not function error

MajorLift committed:
Add missing background method mock

MajorLift committed:
Opt-out `useAssetMetadata`

MajorLift commented:
Hey @davidmurdoch following up on the requested changes!

- "`background[method]` is not a function" error handling:
  - Reverted: https://github.com/MetaMask/metamask-extension/pull/37480/commits/189e390fb7fc480e94b1428f2d3108bf1822901d
  - Fixed breaking unit test: https://github.com/MetaMask/metamask-extension/pull/37480/commits/be6e74934a585f0e215a0bd7653d68831cbf2193
  
- Convoluted `in` check was unnecessary. Reverted: https://github.com/MetaMask/metamask-extension/pull/37480/commits/6788b0f25c55023ad01224d6b657f6d7a9a52a09

- undefined `pendingConfirmations` in confirm-alert files:
  - Reverted and opted-out: https://github.com/MetaMask/metamask-extension/pull/37480/commits/7969aa6d7ecdf64afc6d0b8f0f1d01fa495a4bde
  
- `abortControllerRef` in `useAssetMetadata`:
  - Reverted unnecessary useEffect cleanup: https://github.com/MetaMask/metamask-extension/pull/37480/commits/01094ea2b1a5e03b5fa21b9b4122caa05545ae8e
  - Opted-out `useAssetMetadata`: https://github.com/MetaMask/metamask-extension/pull/37480/commits/dda7ecb115fe012009f1d35fa207392b6dde86f7
    - Reasoning: https://github.com/MetaMask/metamask-extension/pull/37480#discussion_r2561258989

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift committed:
Remove opt-out from root

MajorLift committed:
Merge branch 'main' into jongsun/build/251103-enable-react-compiler

MajorLift commented on ui/hooks/useCurrencyRatePolling.ts:
```diff
@@ -3,7 +3,7 @@ import { useSelector } from 'react-redux';
   getChainIdsToPoll,
   getUseCurrencyRateCheck,
-  useSafeChainsListValidationSelector,
+  getUseSafeChainsListValidation,
```
This is untrue. Both files have been updated.

MajorLift commented on ui/hooks/useMultiPolling.ts:
```diff
@@ -44,19 +59,19 @@ const useMultiPolling = <PollingInput>(
+      prevPollingInputStringified.current = null;
     };
-  }, []);
+  }, [usePollingOptions, hasPollingInputChanged, completedOnboarding]);
```
Any additional effect re-runs impose minimal expense as they are terminated immediately by the early exit. 

The overall overhead is unchanged from the previous implementation, which evaluated `Object.is()` over the complex expression `usePollingOptions.input && JSON.stringify(usePollingOptions.input)` on every render. The execution is just moved to a different stage -- outside of the dependency array.

That said, it's true that this is not optimal. 
TODO: Fix `JSON.stringify()` in deps arrays with `useEqualityCheck` or some other equality fn construct.

MajorLift commented on ui/store/background-connection.ts:
```diff
@@ -30,7 +30,7 @@ export function submitRequestToBackground<R>(
     // the return values for various RPC calls aren't always used. In production
     // builds, this will not happen, and even if it did MM wouldn't work.
-    if (!background) {
+    if (typeof background?.[method] !== 'function') {
```
Resolved: https://github.com/MetaMask/metamask-extension/pull/37480#issuecomment-3577449944

MajorLift commented on ui/hooks/snaps/useSnapNavigation.ts:
```diff
@@ -20,9 +20,9 @@ const useSnapNavigation = () => {
+      handleSnapNavigate,
     }),
-    [useSnapNavigate],
+    [handleSnapNavigate],
```
Ok good find: There are 2 instances of `const { navigate } = useSnapNavigation()`.
However, that code is already broken so fixing it is out-of-scope for this PR. 
We'll address that in a follow-up ticket. 

TODO: Fix undefined `navigate` in `ui/components/app/snaps/snap-ui-{link,markdown}.js`
